src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -263,6 +270,68 @@ public BulkInsertOperation(string database, IDocumentStore store, BulkInsertOpti
 
             _generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(_requestExecutor.Conventions,
                 entity => AsyncHelpers.RunSync(() => _requestExecutor.Conventions.GenerateDocumentIdAsync(database, entity)));
+
+            _streamLock = new SemaphoreSlim(1,1);
+            _lastWriteToStream = DateTime.UtcNow;
+            _timer = new Timer(HandleHeartbeat,
+                new WeakReference(this),
+                HeartbeatCheckInterval,
+                HeartbeatCheckInterval);
+        }
+
+        private static void HandleHeartbeat(object state)
+        {
+            var bulkInsert = (BulkInsertOperation)(((WeakReference)state).Target);
+            if (bulkInsert == null)
+                return;
+            AsyncHelpers.RunSync( () => SendHeartBeat(bulkInsert));
+
+            bulkInsert._timer.Change(HeartbeatCheckInterval, Timeout.InfiniteTimeSpan);
+        }
+
+        private static async Task SendHeartBeat(BulkInsertOperation bulkInsert)
+        {
+            if (DateTime.UtcNow.Ticks - bulkInsert._lastWriteToStream.Ticks >= HeartbeatCheckInterval.Ticks)
+            {
+                try
+                {
+                    await bulkInsert._streamLock.WaitAsync().ConfigureAwait(false);
```

Comment:
shouldn't those methods be done after we check server version?

Reply:
This is there only because if the delay is first and we didn't send anything to the server and we won't have the version yet